RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,358 @@
+immediately. This is done as a performance optimisation to avoid overhead for
+concrete `Tensor` objects.
+
+It will be the job of the dispatcher to extract `Tensor`-like objects from the
+argument list, however, arguments of type `Optional[Tensor]` will be considered
+`Tensor`-like. If one gets a compound or dependent type such as `List[Tensor]`
+or `Tuple[Tensor, ...]` or `Tuple[Tensor, int]`, the dispatcher will have the job
+of extracting an iterable of objects that *could* be `Tensor`-like.
+
+### Generic implementation of `__torch_function__`
+`torch.Tensor` will gain a generic `__torch_function__` of the following form:
+
+```python
+class Tensor:
+    def __torch_function__(self, func, types, args, kwargs):
+        if not all(issubclass(type(self), t) for t in types):
+            return NotImplemented
+        
+        # Defer to internal implementation
+        ret = func._implementation(*args, **kwargs)
+        if type(self) is not Tensor and isinstance(ret, Tensor):
+            ret = Tensor.make_subclass(ret, type(self))
+        return ret
+```
+
+This method has the effect of passing through subclasses through all
+functions/methods as intended.
+
+This corresponds exactly to the implementation `numpy.ndarray` gains in [[4]],
+except for the fact that subclasses are passed through via another internal
+mechanism (namely the `__array_finalize__` protocol) there, as well as the fact
+that we are checking subclassing against `type(self)` instead of `Tensor`. This
+has the side-effect of ensuring unrelated class trees are not merged, which is
+an inconsistency in NumPy's own design. Specifically, consider the example of
+two direct subclasses of `torch.Tensor`. Both will return `NotImplemented`, and
+therefore, the check will fail and `TypeError` will be raised.
+
+Since subclasses are checked before superclasses in `__torch_function__`, it is
+guaranteed that the subclass implementation will be called first. In this
+instance, since `type(self)` is a subclass of all types, the code will
+continue. Since `type(self)` is not `torch.Tensor`, a view into the original
+data is created and returned.
+
+This also works for all operators: `__add__`, `__getitem__` and so on since in
+Python these operators are just dunder methods of the corresponding class.
+
+### The need for `super().__torch_function__`
+To access super, one would do the following:
+```python
+class SubTensor(torch.Tensor):
+    def __torch_function__(self, func, types, args, kwargs):
+        # Pre-processing here
+        val = super().__torch_function__(
+            func,
+            tuple(
+                t for t in types if not issubclass(t, SubTensor)
+            ),
+            args,
+            kwargs
+        )
+        # Post processing here
+```
+
+This way `__torch_function__` knows the list of types to dispatch to, and it
+will _not_ dispatch to `SubTensor` anymore in this example.
```

Comment:
Under my proposal, `__torch_function__` never dispatches to `SubTensor` methods ever; instead, `__torch_function__` is the base case that gets us to call `func._implementation` (terminates `__torch_function__` handling)